x/ibc/callbacks/keeper/keeper.go

```diff
@@ -0,0 +1,429 @@
+	isolatedAddrHex := common.BytesToAddress(isolatedAddr.Bytes())
+
+	acc := k.authKeeper.NewAccountWithAddress(ctx, receiver)
+	k.authKeeper.SetAccount(ctx, acc)
+
+	// Ensure receiver address is equal to the isolated address.
+	if receiverHex.Cmp(isolatedAddrHex) != 0 {
+		return errorsmod.Wrapf(types.ErrInvalidReceiverAddress, "expected %s, got %s", isolatedAddrHex.String(), receiverHex.String())
+	}
+
+	contractAddr := common.HexToAddress(contractAddress)
+	contractAccount := k.evmKeeper.GetAccountOrEmpty(ctx, contractAddr)
+
+	// Check if the contract address contains code.
+	// This check is required because if there is no code, the call will still pass on the EVM side,
+	// but it will ignore the calldata and funds may get stuck.
+	if !contractAccount.IsContract() {
+		return errorsmod.Wrapf(types.ErrContractHasNoCode, "provided contract address is not a contract: %s", contractAddr)
+	}
+
+	// Check if the token pair exists and get the ERC20 contract address
+	// for the native ERC20 or the precompile.
+	// This call fails if the token does not exist or is not registered.
+	token := transfertypes.Token{
+		Denom:  data.Token.Denom,
+		Amount: data.Token.Amount,
+	}
+	coin := ibc.GetReceivedCoin(packet.(channeltypes.Packet), token)
+
+	tokenPairID := k.erc20Keeper.GetTokenPairID(ctx, coin.Denom)
+	tokenPair, found := k.erc20Keeper.GetTokenPair(ctx, tokenPairID)
+	if !found {
+		return errorsmod.Wrapf(types.ErrTokenPairNotFound, "token pair for denom %s not found", data.Token.Denom.IBCDenom())
+	}
+	amountInt, ok := math.NewIntFromString(data.Token.Amount)
+	if !ok {
+		return errorsmod.Wrapf(types.ErrNumberOverflow, "amount overflow")
+	}
+
+	erc20 := contracts.ERC20MinterBurnerDecimalsContract
+
+	remainingGas := math.NewIntFromUint64(cachedCtx.GasMeter().GasRemaining()).BigInt()
+
+	// Call the EVM with the remaining gas as the maximum gas limit.
+	// Up to now, the remaining gas is equal to the callback gas limit set by the user.
+	// NOTE: use the cached ctx for the EVM calls.
+	res, err := k.evmKeeper.CallEVM(cachedCtx, erc20.ABI, receiverHex, tokenPair.GetERC20Contract(), true, remainingGas, "approve", contractAddr, amountInt.BigInt())
+	if err != nil {
+		return errorsmod.Wrapf(types.ErrAllowanceFailed, "failed to set allowance: %v", err)
+	}
+
+	// Consume the actual used gas on the original callback context.
+	ctx.GasMeter().ConsumeGas(res.GasUsed, "callback allowance")
+	remainingGas = remainingGas.Sub(remainingGas, math.NewIntFromUint64(res.GasUsed).BigInt())
+	if ctx.GasMeter().IsOutOfGas() || remainingGas.Cmp(big.NewInt(0)) < 0 {
+		return errorsmod.Wrapf(types.ErrOutOfGas, "out of gas")
+	}
+
+	var allowance *big.Int
+	err = erc20.ABI.UnpackIntoInterface(&allowance, "allowance", res.Ret)
+	if err != nil {
+		return errorsmod.Wrapf(types.ErrAllowanceFailed, "failed to unpack allowance: %v", err)
+	}
+
+	if allowance.Cmp(big.NewInt(1)) != 0 {
+		return errorsmod.Wrapf(types.ErrAllowanceFailed, "failed to set allowance")
+	}
```

Comment:
I think this is really confusing to read. It's not clear to me that the `allowance` value being returned here is actually a boolean for success and not the actual allowance value being set by the `approve` call.

Are we not able to actually unpack to a bool here instead?